include/zephyr/drivers/spi.h

```diff
@@ -138,6 +138,28 @@ extern "C" {
 #define SPI_LINES_MASK		(0x3U << 16)
 /** @} */
 
+#ifdef CONFIG_SPI_EXTENDED_MODES
+/**
+ * @name SPI packet size (in units of frames)
+ * @{
+ *
+ */
+#define SPI_PACKET_SIZE_SHIFT (18U)
+#define SPI_PACKET_SIZE_MASK (0xFU << SPI_PACKET_SIZE_SHIFT)
+
+/* Offset the packet size so if it isn't specified (val. 0), its value is 1 */
+#define SPI_PACKET_SIZE_OFFSET (1U)
+
+#define SPI_PACKET_SIZE_GET(_operation_)                                        \
+	((((_operation_) & SPI_PACKET_SIZE_MASK) >> SPI_PACKET_SIZE_SHIFT)          \
+	    + SPI_PACKET_SIZE_OFFSET)
+
+#define SPI_PACKET_SIZE_SET(_packet_size_)                                      \
+		(((_packet_size_) - SPI_PACKET_SIZE_OFFSET) <<                          \
+			SPI_PACKET_SIZE_SHIFT)
+/** @} */
+#endif /* CONFIG_SPI_EXTENDED_MODES */
```

Comment:
What would be better place to put this instead? The `spi_ll_stm32.h` header is private. What is the standard Zephyr approach for platform-specific definitions in a driver?

Reply:
About the depth of the FIFO, it would be good to add a new parameter to indicate this to `st,stm32-spi-fifo.yaml`, so the driver can return an error if the FIFO threshold (or packet size as it stands now in this PR) set by the user is too big. I'm happy to do this within this PR if you agree.

Reply:
I have taken a quick look at the current Zephyr's drivers to find out more about this. For instance, the ADC driver has a directory `include/zephyr/drivers/adc` which contains (among others) the 2 following header files:

 * lmp90xxx.h
 * ads114s0x.h

These seem to be device-specific headers. Maybe I could apply the same principle to this PR, that is, to move the `SPI_PACKET_SIZE` definitions to a file like `include/zephyr/drivers/spi/stm32.h`.

Reply:
Headers that end up in include/zephyr/drivers/<api>/ are about drivers that let some features or configuration options to be manipulated by the user program.

In your case, you don't need to expose anything publicly. Actually, adding a new parameter to the dts binding as you noticed, would be the way to go.

Reply:
Notice that I referred specifically to the FIFO depth to be added to device tree. The FIFO threshold shouldn't be added to devicetree: it might vary depending on the length of the data to be sent, which is provided in the SPI configuration struct. For example, if the FIFO threshold is set as 8 bytes, a transfers of size 9 bytes will lock for ever. This is because RXP is set when at least 1 packet is available in the RxFIFO, meaning that the last byte sent won't make RXP to be set, hence preventing the SPI driver to read the last byte of the transmission.

Reply:
@tbursztyka this was discussed in the dev meeting and it was approved that the FIFO depth should be in devicetree and the FIFO threshold/packet size should be provided along with the other spi config. parameters. Do you see any problem with this?

Reply:
Packet size terminology seems very specific to st here. NXP's LPSPI uses the term frame size for I believe something similar. But really shouldn't all this be done internally by the driver to optimally send whatever data is there? Maybe I missed something

Reply:
ST uses frame size as well, but to define how the FIFO is addressed (8, 16, 32 bits). Packet size is given in units of frames. Thus, a packet size of 3 is 3 bytes if the frame size is 8 and it's 12 bytes if frame size is 32 bits. In any case, I'm happy to name this FIFO threshold instead.

It's true that an 'optimum' (biggest possible) FIFO threshold can be inferred from the FIFO depth (defined in DT) and the transfer total length, but this has 2 drawbacks:
 * It doesn't give the user the option to choose the desired FIFO threshold. It could happen that the biggest possible FIFO threshold doesn't work because the target SPI device cannot cope with it.
 * Testing becomes more confusing and difficult. It's very important to test this feature intensively as it can cause the SPI peripheral to block for ever if misused. I want to remark that test cases become way more difficult to understand. 

I had that implementation before and decided to change it to this.

Reply:
Kind reminder to @tbursztyka and @teburd to clarify this point please. After all the explanation above, there are 2 options:
 * Provide the FIFO threshold as a parameter
 * Calculate the optimal FIFO threshold inside the driver automatically.

I strongly suggest the first option, putting the required macros in a H7 SPI specific header file as explained in another comment. Please could you confirm if you agree?

Reply:
I guess the question is how does this play out in an independent driver for a SPI device, like a sensor driver going to read a 2kb FIFO over SPI.

In this case, to get the ideal performance it seems like for this particular platform some ifdefry would be needed. That to me seems like a good indicator that exposing these details is the wrong approach but I could be wrong.

Reply:
> I guess the question is how does this play out in an independent driver for a SPI device, like a sensor driver going to read a 2kb FIFO over SPI.

With the current implementation, the FIFO threshold defaults to 1, so other modules using an SPI driver won't notice any change. Possibly they will still notice an improvement in performance.

> In this case, to get the ideal performance it seems like for this particular platform some ifdefry would be needed. 

Yes, a module using this SPI driver would need to check (potentially with an #ifdef) if it is an H7 and then provide a specific FIFO threshold for the particular length it's trying to send.

If this is a problem, I believe (I would need to test it though) that it would be possible to not provide a FIFO threshold in any way or form. This would mean that H7 would always work with FIFO threshold = 1, but it would still use the FIFO capabilities, at least after applying the feedback below, where data is written/read to/from the FIFO while TXP/RXP allows it.

Reply:
I have removed the use of packet size/FIFO threshold.

Reply:
@teburd @tbursztyka unless you have any objection, I will resolve this conversation.